Cython/Compiler/Nodes.py

```diff
@@ -5310,6 +5307,16 @@ def analyse_declarations(self, env):
         self.scope = scope = self.entry.type.scope
         if scope is not None:
             scope.directives = env.directives
+            if "dataclasses.dataclass" in env.directives:
+                is_frozen = False
+                dataclass_directives = env.directives["dataclasses.dataclass"]
+                if dataclass_directives:
+                    frozen_directive = dataclass_directives[1].get('frozen', None)
+                    is_frozen = frozen_directive and frozen_directive.is_literal and frozen_directive.value
+                if is_frozen:
+                    scope.is_dataclass = "frozen"
+                else:
+                    scope.is_dataclass = True
```

Comment:
I really like conditional expressions for cases like this because they underline the intention to set a variable or attribute, just in different ways depending on a condition.
```suggestion
                scope.is_dataclass = "frozen" if is_frozen else True
```